gcs/smart_entry.hh

```diff
@@ -0,0 +1,130 @@
+#ifndef GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_SMART_ENTRY_HH
+#define GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_SMART_ENTRY_HH
+
+#include <variant>
+#include <vector>
+#include <gcs/variable_id.hh>
+#include <gcs/integer.hh>
+#include <algorithm> //std::sort
+using namespace gcs;
+
```

Comment:
All of this should be in `namespace gcs`. I think it might even be worth having a sub-namespace for some of it, so that shorter class names can be used?